### PR TITLE
Improve pixi drawing performance by caching off as much primitive drawing to a single Graphics object as possible

### PR DESCRIPTION
**Pull request by tlein**

### CODE REVIEW

File: src/duckling/canvas/drawing/entity-drawer.service.ts [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
 
 import {RenderPriorityService} from './render-priority.service';
 import {DrawnConstruct, setConstructPosition} from './drawn-construct';
 
-/**
- * Function type used to draw attributes.
- */
-export type AttributeDrawer<T> = (attribute : T, assetService? : any) => DrawnConstruct;
+export type AttributeDrawer<T> = {
+    paint?(attribute : T, position : Vector, graphics : Graphics) : void,
```
Comment (tlein) [inline src/duckling/canvas/drawing/entity-drawer.service.ts]: This is weird syntax: 

`paint? : (attribute : T, position : Vector, graphics : Graphics) => void`

File: src/duckling/canvas/drawing/entity-drawer.service.ts [hunk 2]
```diff
@@ -70,49 +60,79 @@ export class EntityDrawerService extends BaseAttributeService<AttributeDrawer<At
...
+                drawnConstructs.push(construct);
+            }
+            newCache[entity.key] = entry;
+        }
+        this._cache = newCache;
+        return drawnConstructs;
+    }
+
+    private _paintEntitySystem(entities : TaggedEntity[], graphics : Graphics) {
+        graphics.clear();
```
Comment (tlein) [inline src/duckling/canvas/drawing/entity-drawer.service.ts]: This clear should only be in the map-editor when the painted cache is killed

**Comment by jeffswenson** [comment]
Instead of forking the drawing infrastructure you should create a Painter DrawnConstruct.

File: src/duckling/canvas/drawing/entity-drawer.service.ts [hunk 3]
```diff
@@ -70,49 +60,79 @@ export class EntityDrawerService extends BaseAttributeService<AttributeDrawer<At
...
+        let sortedEntities = this._renderPriority.sortEntities(entitySystem);
+        if (graphics) {
+            this._paintEntitySystem(sortedEntities, graphics);
+        }
+        return this._getEntitySystemDrawables(sortedEntities);
+    }
+
+    private _getEntitySystemDrawables(entities : TaggedEntity[]) : DrawnConstruct[] {
+        let drawnConstructs : DrawnConstruct[] = [];
+        let newCache : EntityCache = {};
```
Comment (jeffswenson) [inline src/duckling/canvas/drawing/entity-drawer.service.ts]: I think we should have the cache held by the MapEditor and this should expose a InvalidateDrawableCache observer that fires an event when something causes the drawable cache to be stale.

File: src/duckling/canvas/map-editor.component.ts [hunk 4]
```diff
@@ -199,98 +172,146 @@ export class MapEditorComponent implements AfterViewInit, OnDestroy {
...
+        await this.projectService.openMap(mapKey);
+        this._markAllCachesDirty();
     }
 
     private _redrawAllDisplayObjects() {
-        this._canvasBackgroundDisplayObject = this._buildCanvasBackground();
-        this._canvasBorderDisplayObject = this._buildCanvasBorder();
-        this._gridDisplayObject = this._buildGrid();
-        this._toolDisplayObject = this.tool.getDisplayObject(this.scale);
+        if (this._drawableCache[BACKGROUND_CACHE_KEY].dirty) {
```
Comment (jeffswenson) [inline src/duckling/canvas/map-editor.component.ts]: There should be a way to dry this up.

Comment (jeffswenson) [reply src/duckling/canvas/map-editor.component.ts]: I think DrawableCache should be a class. The constructor should take a function that is called to rebuild the cache. You ould transform things like:

this._markeCacheDirty(cacheKey) to this.cache.cacheKey.dirty = true;

File: src/duckling/canvas/drawing/entity-drawer.service.ts [hunk 5]
```diff
@@ -23,36 +23,32 @@ interface EntityCacheEntry {
...
-     * Get a DisplayObject for the attribute.
-     * @param key Attribute key of the attribute that should be drawn.
-     * @param attribute Attribute that needs to be drawn.
-     * @return A DrawnConstruct that describes how an entity is drawn.
-     */
-    drawAttribute(key : AttributeKey, entity : Entity) : DrawnConstruct {
+    getAttributeDrawable(key : AttributeKey, entity : Entity) : DrawnConstruct {
         let drawer = this.getImplementation(key);
-        if (drawer) {
+        if (drawer && drawer) {
```
Comment (jeffswenson) [inline src/duckling/canvas/drawing/entity-drawer.service.ts]: drawer && drawer

File: src/duckling/canvas/drawing/drawn-construct.ts [hunk 6]
```diff
@@ -1,13 +1,9 @@
-import {Container, DisplayObject, Point} from 'pixi.js';
+import {Container, DisplayObject, Point, Graphics} from 'pixi.js';
```
Comment (jeffswenson) [inline src/duckling/canvas/drawing/drawn-construct.ts:1]: As we discussed offline
I think DrawnConstruct should be a polymorphic interface with drawable?, paint?, and layer:string properties.

The loop used to draw entites should be replaced with
for (entity) for (attribute) constructs.push(getConstruct(attribute))
constructs.sort(construct.layer)

When drawing, you should create a new graphics object for each layer.

**Review by jeffswenson — CHANGES_REQUESTED** [review]

File: src/duckling/canvas/drawing/entity-drawer.service.ts (hunk 5 above)
Comment (tlein) [reply src/duckling/canvas/drawing/entity-drawer.service.ts]: fixed

File: src/duckling/canvas/drawing/drawn-construct.ts (hunk 6 above)
Comment (tlein) [reply src/duckling/canvas/drawing/drawn-construct.ts:1]: fixed

File: src/duckling/canvas/drawing/entity-drawer.service.ts (hunk 3 above)
Comment (tlein) [reply src/duckling/canvas/drawing/entity-drawer.service.ts]: fixed

File: src/duckling/canvas/drawing/drawn-construct.ts [hunk 7]
```diff
@@ -1,138 +1,84 @@
...
-    } else {
-        throw Error("Unknown DrawnConstruct type in drawn-construct::displayObjectForDrawnConstruct");
+    protected _applyDisplayObjectTransforms(displayObject : DisplayObject) {
+        if (this.transformProperties.anchor.x || this.transformProperties.anchor.y) {
+            let bounds = this._displayObjectBounds(displayObject);
+            displayObject.pivot.x = bounds.dimension.x * this.transformProperties.anchor.x;
+            displayObject.pivot.y = bounds.dimension.y * this.transformProperties.anchor.y;
+        }
+
+        if (this.transformProperties.rotation) {
```
Comment (jeffswenson) [inline src/duckling/canvas/drawing/drawn-construct.ts]: this optimization results in a problem if you change rotation from a non-zero value to a zero value.

**Review by jeffswenson — APPROVED** [review]

**Pull request merged**